src/resources.rs

```diff
@@ -159,20 +163,51 @@ impl Resource {
     let mut table = RESOURCE_TABLE.lock().unwrap();
     let maybe_repr = table.get_mut(&self.rid);
     match maybe_repr {
-      None => panic!("bad rid"),
+      None => Err(std::io::Error::new(
+        std::io::ErrorKind::Other,
+        "Listener has been closed",
+      )),
       Some(repr) => match repr {
-        Repr::TcpListener(ref mut s) => s.poll_accept(),
+        Repr::TcpListener(ref mut s, _) => s.poll_accept(),
         _ => panic!("Cannot accept"),
       },
     }
   }
 
+  /// Track pending tasks with task_id as identifier (for TcpListener resource)
+  /// Insert the task into map for tracking
+  pub fn track_task(&mut self, task_id: usize) {
+    let mut table = RESOURCE_TABLE.lock().unwrap();
+    // Only register if is TcpListener
+    if let Some(Repr::TcpListener(_, m)) = table.get_mut(&self.rid) {
+      m.insert(task_id, futures::task::current());
+    }
+  }
+
+  /// Remove a task from tracking (for TcpListener resource)
+  /// Happens when the task is done and thus no further tracking is needed
+  pub fn untrack_task(&mut self, task_id: usize) {
+    let mut table = RESOURCE_TABLE.lock().unwrap();
+    if let Some(Repr::TcpListener(_, m)) = table.get_mut(&self.rid) {
+      m.remove(&task_id);
+    }
+  }
+
   // close(2) is done by dropping the value. Therefore we just need to remove
   // the resource from the RESOURCE_TABLE.
   pub fn close(&self) {
     let mut table = RESOURCE_TABLE.lock().unwrap();
     let r = table.remove(&self.rid);
     assert!(r.is_some());
+    // If TcpListener, we must kill all pending accepts!
+    if let Repr::TcpListener(l, m) = r.unwrap() {
+      // Drop first
+      std::mem::drop(l);
```

Comment:
I don’t think this is necessary - it will be dropped with the above remove

Reply:
Indeed... Forgot to remove this (was part of initial experiment...)